Client/trunk/ParaEngineClient/3dengine/GeosetObject.cpp

```diff
@@ -0,0 +1,271 @@
+				// block light
+				float fBlockLightness=Math::Max(pBlockWorldClient->GetLightBrightnessLinearFloat(brightness[0]),0.1f);
+				sceneState->GetCurrentLightStrength().y=fBlockLightness;
+				// sun light
+				float fSunLightness=Math::Max(pBlockWorldClient->GetLightBrightnessLinearFloat(brightness[1]),0.1f);
+				sceneState->GetCurrentLightStrength().x=fSunLightness;
+
+				float fLightness=Math::Max(fBlockLightness,fSunLightness*pBlockWorldClient->GetSunIntensity());
+				//if(m_fLastBlockLight!=fLightness)
+				//{
+				//	float fMaxStep=(float)(sceneState->dTimeDelta*0.5f);
+				//	if(dwPositionHash==m_dwLastBlockHash||m_dwLastBlockHash==0)
+				//		m_fLastBlockLight=fLightness;
+				//	else
+				//		Math::SmoothMoveFloat1(m_fLastBlockLight,fLightness,fMaxStep);
+
+				//	fLightness=m_fLastBlockLight;
+				//}
+				//else
+				//{
+				//	m_dwLastBlockHash=dwPositionHash;
+				//}
+				sceneState->GetLocalMaterial().Ambient=(LinearColor(fLightness*0.7f,fLightness*0.7f,fLightness*0.7f,1.f));
+				sceneState->GetLocalMaterial().Diffuse=(LinearColor(fLightness*0.4f,fLightness*0.4f,fLightness*0.4f,1.f));
+
+				sceneState->EnableLocalMaterial(true);
+
+				bUsePointTextureFilter=bUsePointTextureFilter||pBlockWorldClient->GetUsePointTextureFiltering();
+			}
+		}
+
+		if(bUsePointTextureFilter)
+		{
+			pEffectManager->SetSamplerState(0,D3DSAMP_MINFILTER,D3DTEXF_POINT);
+			pEffectManager->SetSamplerState(0,D3DSAMP_MAGFILTER,D3DTEXF_POINT);
+		}
+		else
+		{
+			pEffectManager->SetSamplerState(0,D3DSAMP_MINFILTER,pEffectManager->GetDefaultSamplerState(0,D3DSAMP_MINFILTER));
+			pEffectManager->SetSamplerState(0,D3DSAMP_MAGFILTER,pEffectManager->GetDefaultSamplerState(0,D3DSAMP_MAGFILTER));
+		}
+
+		CApplyObjectLevelParamBlock p(mParent->GetEffectParamBlock());
+		_draw(sceneState,mat,p.GetParamsBlock());
+
+		sceneState->EnableLocalMaterial(false);
+	}
+
+	return S_OK;
+}
+
+void ParaEngine::CGeosetObject::_draw(SceneState * sceneState,Matrix4 * mxWorld,CParameterBlock* params)
+{
+	auto* pAI=static_cast<CParaXAnimInstance*>(mParent->GetAnimInstance());
+	Matrix4 mat;
+	if(pAI->UpdateWorldTransform(sceneState,mat,*mxWorld))
+	{
+		CGlobals::GetWorldMatrixStack().push(mat);
+		// draw model
+		RenderDevicePtr  pd3dDevice=CGlobals::GetRenderDevice();
+
+		float fCameraToObjectDist=sceneState->GetCameraToCurObjectDistance();
+		// scale the model
+		Matrix4 mat2;
+		mat2=CGlobals::GetWorldMatrixStack().SafeGetTop();
+		CGlobals::GetWorldMatrixStack().push(mat2);
+
+		// draw this model
+		int nIndex=(sceneState && sceneState->IsLODEnabled())?mEntity->GetLodIndex(fCameraToObjectDist):0;
+		CParaXModel* pModel=mEntity->GetModel(nIndex);
+
+		if(pModel)
+		{
+			pModel->m_CurrentAnim=pAI->m_CurrentAnim;
+			pModel->m_NextAnim=pAI->m_NextAnim;
+			pModel->m_BlendingAnim=pAI->m_BlendingAnim;
+			pModel->blendingFactor=pAI->m_fBlendingFactor;
+			pModel->animate(sceneState,nullptr,pAI);
```

Comment:
bone mapping and animation data sharing is not implemented yet. try to implement by changing the pModel interface, etc.